File: typescript/typelayer-common/bin/encodings/msgpack.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Module: encodings.msgpack
const typelayerCommonDynamic = __importStar(require("../dynamic"));
const typelayerCommonTypes = __importStar(require("../types"));
class MsgPackEncoding {
    constructor(arg) {
        this.kind = "MsgPackEncoding";
        Object.assign(this, arg);
    }
    encodingLabel() {
        return "msgpack";
    }
    encodingMimetype() {
        return "application/msgpack";
    }
    encodingStartDecoding(val) {
        return new MsgPackDecoder({
            index: 0,
            reader: new typelayerCommonTypes.ArrayReader(val)
        });
    }
    encodingStartEncoding() {
        return new MsgPackEncoder({
            writer: new typelayerCommonTypes.ArrayWriter(0)
        });
    }
}
exports.MsgPackEncoding = MsgPackEncoding;
class MsgPackDecoder {
    constructor(arg) {
        this.kind = "MsgPackDecoder";
        Object.assign(this, arg);
    }
    decoderPath() {
        return `${this.index}`;
    }
    decoderPopBool() {
        let headByte;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xc2:
                return true;
                break;
            case 0xc3:
                return false;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected bool", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopDouble() {
        let headByte;
        let value;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xcb:
                value = this.reader.view.getFloat64(this.index);
                this.index = 8 + this.index;
                return value;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected double", { path: this.decoderPath() });
        }
    }
    decoderPopField() {
        let fieldSize;
        let fieldValue;
        let headByte;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xc6:
                fieldSize =
                    this.reader.array[this.index + 1] |
                        (this.reader.array[this.index] << 16);
                this.index = 2 + this.index;
                fieldValue = this.reader.getString(this.index, fieldSize);
                this.index = fieldSize + this.index;
                return fieldValue;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected field", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopHasMoreList() {
        return 0;
    }
    decoderPopHasMoreMap() {
        return 0;
    }
    decoderPopHasMoreStruct() {
        return 0;
    }
    decoderPopInteger() {
        let headByte;
        let value;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xcf:
                value = this.reader.getInt64(this.index);
                this.index = 8 + this.index;
                return value;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected integer", { path: this.decoderPath() });
        }
    }
    decoderPopKey() {
        let headByte;
        let keySize;
        let keyValue;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xdb:
                keySize =
                    this.reader.array[this.index + 1] |
                        (this.reader.array[this.index] << 16);
                this.index = 2 + this.index;
                keyValue = this.reader.getString(this.index, keySize);
                this.index = keySize + this.index;
                return keyValue;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected key", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopList() {
        let headByte;
        let value;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xdd:
                value = this.reader.view.getUint32(this.index);
                this.index = 4 + this.index;
                return value;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected list", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopMap() {
        let headByte;
        let value;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xdf:
                value = this.reader.view.getUint32(this.index);
                this.index = 4 + this.index;
                return value;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected map", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopMaybe() {
        let headByte;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xc0:
                return false;
                break;
            case 0x91:
                return true;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected maybe", {
                    path: this.decoderPath()
                });
        }
    }
    decoderPopString() {
        let headByte;
        let strSize;
        let strValue;
        headByte = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (headByte) {
            case 0xdb:
                strSize = this.reader.view.getUint32(this.index);
                this.index = 4 + this.index;
                strValue = this.reader.getString(this.index, strSize);
                this.index = strSize + this.index;
                return strValue;
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected string", { path: this.decoderPath() });
        }
    }
    decoderPopStruct(defaultConstructor) {
        let cName;
        let cSize;
        let numFields;
        let value;
        value = this.reader.array[this.index];
        this.index = 1 + this.index;
        switch (value) {
            case 0xde:
                numFields = this.reader.view.getUint32(this.index);
                this.index = 4 + this.index;
                if (0 < numFields) {
                    if (this.reader.array[this.index] === 0xc4) {
                        this.index = 4 + this.index;
                        cSize = this.reader.view.getUint32(this.index);
                        this.index = 4 + this.index;
                        cName = this.reader.getString(this.index, cSize);
                        this.index = cSize + this.index;
                        return [cName, numFields - 1];
                    }
                    else {
                        return [defaultConstructor, numFields];
                    }
                }
                else {
                    return [defaultConstructor, numFields];
                }
                break;
            default:
                throw new typelayerCommonDynamic.DynamicGetException("Expected struct", { path: this.decoderPath() });
        }
    }
}
exports.MsgPackDecoder = MsgPackDecoder;
class MsgPackEncoder {
    constructor(arg) {
        this.kind = "MsgPackEncoder";
        Object.assign(this, arg);
    }
    encoderPutBool(value) {
        let newWriter;
        if (value) {
            this.writer.reserve(1);
            this.writer.array[this.writer.offset] = 0xc2;
            this.writer.offset += 1;
            newWriter = this.writer;
        }
        else {
            this.writer.reserve(1);
            this.writer.array[this.writer.offset] = 0xc3;
            this.writer.offset += 1;
            newWriter = this.writer;
        }
        this.writer = newWriter;
        return;
    }
    encoderPutDouble(value) {
        let newWriter;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xcb;
        this.writer.offset += 1;
        this.writer.reserve(8);
        this.writer.view.setFloat64(this.writer.offset, value);
        this.writer.offset += 8;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutElementEnd(value) {
        let newWriter;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutField(value) {
        let newWriter;
        let strAsBytes;
        strAsBytes = typelayerCommonTypes.stringToBytes(value);
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xc6;
        this.writer.offset += 1;
        this.writer.reserve(2);
        this.writer.view.setUint16(this.writer.offset, strAsBytes.length);
        this.writer.offset += 2;
        this.writer.reserve(strAsBytes.length);
        this.writer.array.set(strAsBytes, this.writer.offset);
        this.writer.offset += strAsBytes.length;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutFieldEnd(value) {
        let newWriter;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutInteger(value) {
        let newWriter;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xcf;
        this.writer.offset += 1;
        this.writer.setInt64(this.writer.offset, value);
        this.writer.offset += 8;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutKey(value) {
        let newWriter;
        let strAsBytes;
        strAsBytes = typelayerCommonTypes.stringToBytes(value);
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xdb;
        this.writer.offset += 1;
        this.writer.reserve(2);
        this.writer.view.setUint16(this.writer.offset, strAsBytes.length);
        this.writer.offset += 2;
        this.writer.reserve(strAsBytes.length);
        this.writer.array.set(strAsBytes, this.writer.offset);
        this.writer.offset += strAsBytes.length;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutList(value) {
        let newWriter;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xdd;
        this.writer.offset += 1;
        this.writer.reserve(4);
        this.writer.view.setInt32(this.writer.offset, value);
        this.writer.offset += 4;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutMap(value) {
        let newWriter;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xdf;
        this.writer.offset += 1;
        this.writer.reserve(4);
        this.writer.view.setInt32(this.writer.offset, value);
        this.writer.offset += 4;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutMaybe(value) {
        let newWriter;
        if (value) {
            this.writer.reserve(1);
            this.writer.array[this.writer.offset] = 0x91;
            this.writer.offset += 1;
            newWriter = this.writer;
        }
        else {
            this.writer.reserve(1);
            this.writer.array[this.writer.offset] = 0xc0;
            this.writer.offset += 1;
            newWriter = this.writer;
        }
        this.writer = newWriter;
        return;
    }
    encoderPutString(value) {
        let newWriter;
        let strAsBytes;
        strAsBytes = typelayerCommonTypes.stringToBytes(value);
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xdb;
        this.writer.offset += 1;
        this.writer.reserve(4);
        this.writer.view.setInt32(this.writer.offset, strAsBytes.length);
        this.writer.offset += 4;
        this.writer.reserve(strAsBytes.length);
        this.writer.array.set(strAsBytes, this.writer.offset);
        this.writer.offset += strAsBytes.length;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutStruct(constructor, numFields) {
        let cAsBytes;
        let newWriter;
        cAsBytes = typelayerCommonTypes.stringToBytes(constructor);
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xde;
        this.writer.offset += 1;
        this.writer.reserve(4);
        this.writer.view.setInt32(this.writer.offset, numFields + 1);
        this.writer.offset += 4;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xc4;
        this.writer.offset += 1;
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0x1;
        this.writer.offset += 1;
        this.writer.offset += this.writer.setString("_", this.writer.offset);
        this.writer.reserve(1);
        this.writer.array[this.writer.offset] = 0xdb;
        this.writer.offset += 1;
        this.writer.reserve(4);
        this.writer.view.setInt32(this.writer.offset, cAsBytes.length);
        this.writer.offset += 4;
        this.writer.reserve(cAsBytes.length);
        this.writer.array.set(cAsBytes, this.writer.offset);
        this.writer.offset += cAsBytes.length;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderPutValueEnd(value) {
        let newWriter;
        newWriter = this.writer;
        this.writer = newWriter;
        return;
    }
    encoderToBinary() {
        return this.writer.finalize();
    }
}
exports.MsgPackEncoder = MsgPackEncoder;

File: python/typelayer_common/typelayer_common/encodings/msgpack.py
from .. import dynamic as typelayer_common_dynamic
from .. import types as typelayer_common_types
from dataclasses import dataclass
from typing import List, Dict, Optional


@dataclass
class MsgPackEncoding(typelayer_common_dynamic.Encoding):

    def encoding_label(self: "MsgPackEncoding") -> str:
    
        return "msgpack"
    def encoding_mimetype(self: "MsgPackEncoding") -> str:
    
        return "application/msgpack"
    def encoding_start_decoding(self: "MsgPackEncoding", val: bytes) -> typelayer_common_dynamic.Decoder:
    
        return MsgPackDecoder(index=0, reader=typelayer_common_types.MemoryViewReader(val))
    def encoding_start_encoding(self: "MsgPackEncoding") -> typelayer_common_dynamic.Encoder:
    
        return MsgPackEncoder(writer=typelayer_common_types.MemoryViewWriter(0))

@dataclass
class MsgPackDecoder(typelayer_common_dynamic.Decoder):
    index: int;
    reader: typelayer_common_types.MemoryViewReader;
    def decoder_path(self: "MsgPackDecoder") -> str:
    
        return f"{self.index}"
    def decoder_pop_bool(self: "MsgPackDecoder") -> bool:
    
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xc2)):
            return True
        elif (head_byte == (0xc3)):
            return False
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected bool", {"path": self.decoder_path()});
    def decoder_pop_double(self: "MsgPackDecoder") -> float:
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xcb)):
            value = self.reader.getFloat64(self.index)
            self.index = (8 + self.index);
            
            return value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected double", {"path": self.decoder_path()});
    def decoder_pop_field(self: "MsgPackDecoder") -> str:
        
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xc6)):
            field_size = (self.reader.getChar((self.index + 1)) | (self.reader.getChar(self.index) << 16))
            self.index = (2 + self.index);
            
            field_value = self.reader.getString(self.index, field_size)
            self.index = (field_size + self.index);
            
            return field_value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected field", {"path": self.decoder_path()});
    def decoder_pop_has_more_list(self: "MsgPackDecoder") -> int:
    
        return 0
    def decoder_pop_has_more_map(self: "MsgPackDecoder") -> int:
    
        return 0
    def decoder_pop_has_more_struct(self: "MsgPackDecoder") -> int:
    
        return 0
    def decoder_pop_integer(self: "MsgPackDecoder") -> int:
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xcf)):
            value = self.reader.getInt64(self.index)
            self.index = (8 + self.index);
            
            return value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected integer", {"path": self.decoder_path()});
    def decoder_pop_key(self: "MsgPackDecoder") -> str:
        
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xdb)):
            key_size = (self.reader.getChar((self.index + 1)) | (self.reader.getChar(self.index) << 16))
            self.index = (2 + self.index);
            
            key_value = self.reader.getString(self.index, key_size)
            self.index = (key_size + self.index);
            
            return key_value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected key", {"path": self.decoder_path()});
    def decoder_pop_list(self: "MsgPackDecoder") -> int:
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xdd)):
            value = self.reader.getUint32(self.index)
            self.index = (4 + self.index);
            
            return value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected list", {"path": self.decoder_path()});
    def decoder_pop_map(self: "MsgPackDecoder") -> int:
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xdf)):
            value = self.reader.getUint32(self.index)
            self.index = (4 + self.index);
            
            return value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected map", {"path": self.decoder_path()});
    def decoder_pop_maybe(self: "MsgPackDecoder") -> bool:
    
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xc0)):
            return False
        elif (head_byte == (0x91)):
            return True
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected maybe", {"path": self.decoder_path()});
    def decoder_pop_string(self: "MsgPackDecoder") -> str:
        
        
        
        head_byte = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (head_byte == (0xdb)):
            str_size = self.reader.getUint32(self.index)
            self.index = (4 + self.index);
            
            str_value = self.reader.getString(self.index, str_size)
            self.index = (str_size + self.index);
            
            return str_value
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected string", {"path": self.decoder_path()});
    def decoder_pop_struct(self: "MsgPackDecoder", default_constructor: str) -> (str, int):
        
        
        
        
        value = self.reader.getChar(self.index)
        self.index = (1 + self.index);
        
        if (value == (0xde)):
            num_fields = self.reader.getUint32(self.index)
            self.index = (4 + self.index);
            
            if ((0 < num_fields)):
                if ((self.reader.getChar(self.index) == (0xc4))):
                    self.index = (4 + self.index);
                    
                    c_size = self.reader.getUint32(self.index)
                    self.index = (4 + self.index);
                    
                    c_name = self.reader.getString(self.index, c_size)
                    self.index = (c_size + self.index);
                    
                    return (c_name, (num_fields - 1))
                else:
                    return (default_constructor, num_fields)
            else:
                return (default_constructor, num_fields)
        else:
                raise typelayer_common_dynamic.DynamicGetException("Expected struct", {"path": self.decoder_path()});

@dataclass
class MsgPackEncoder(typelayer_common_dynamic.Encoder):
    writer: typelayer_common_types.MemoryViewWriter;
    def encoder_put_bool(self: "MsgPackEncoder", value: bool) -> None:
    
        if (value):
            self.writer.setChar((0xc2))
            new_writer = self.writer
        else:
            self.writer.setChar((0xc3))
            new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_double(self: "MsgPackEncoder", value: float) -> None:
    
        self.writer.setChar((0xcb))
        self.writer.setFloat64(value)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_element_end(self: "MsgPackEncoder", value: bool) -> None:
    
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_field(self: "MsgPackEncoder", value: str) -> None:
        
        
        str_as_bytes = (value.encode("utf8"))
        self.writer.setChar((0xc6))
        self.writer.setUint16(len(str_as_bytes))
        self.writer.setBytes(str_as_bytes)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_field_end(self: "MsgPackEncoder", value: bool) -> None:
    
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_integer(self: "MsgPackEncoder", value: int) -> None:
    
        self.writer.setChar((0xcf))
        self.writer.setInt64(value)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_key(self: "MsgPackEncoder", value: str) -> None:
        
        
        str_as_bytes = (value.encode("utf8"))
        self.writer.setChar((0xdb))
        self.writer.setUint16(len(str_as_bytes))
        self.writer.setBytes(str_as_bytes)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_list(self: "MsgPackEncoder", value: int) -> None:
    
        self.writer.setChar((0xdd))
        self.writer.setUint32(value)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_map(self: "MsgPackEncoder", value: int) -> None:
    
        self.writer.setChar((0xdf))
        self.writer.setUint32(value)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_maybe(self: "MsgPackEncoder", value: bool) -> None:
    
        if (value):
            self.writer.setChar((0x91))
            new_writer = self.writer
        else:
            self.writer.setChar((0xc0))
            new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_string(self: "MsgPackEncoder", value: str) -> None:
        
        
        str_as_bytes = (value.encode("utf8"))
        self.writer.setChar((0xdb))
        self.writer.setUint32(len(str_as_bytes))
        self.writer.setBytes(str_as_bytes)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_struct(self: "MsgPackEncoder", constructor: str, num_fields: int) -> None:
        
        
        c_as_bytes = (constructor.encode("utf8"))
        self.writer.setChar((0xde))
        self.writer.setUint32((num_fields + 1))
        self.writer.setChar((0xc4))
        self.writer.setChar((0x1))
        self.writer.setString("_")
        self.writer.setChar((0xdb))
        self.writer.setUint32(len(c_as_bytes))
        self.writer.setBytes(c_as_bytes)
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_put_value_end(self: "MsgPackEncoder", value: bool) -> None:
    
        new_writer = self.writer
        self.writer = new_writer;
        
        return;
    def encoder_to_binary(self: "MsgPackEncoder") -> bytes:
    
        return self.writer.finalize()